components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,297 @@
+    SingleFileArchiveHeader header{
+            .magic{},
+            .version = cArchiveVersion,
+            .metadata_size = packed_metadata_size,
+            .unused{}
+    };
+    std::memcpy(&header.magic, cUnstructuredSfaMagicNumber.data(), sizeof(header.magic));
+    archive_writer.write(reinterpret_cast<char const*>(&header), sizeof(header));
+}
+
+auto write_archive_metadata(FileWriter& archive_writer, std::stringstream const& packed_metadata)
+        -> void {
+    archive_writer.write(packed_metadata.str().data(), packed_metadata.str().size());
+}
+
+auto write_archive_file(std::filesystem::path const& file_path, FileWriter& archive_writer)
+        -> void {
+    FileReader reader(file_path.string());
+    std::array<char, cReadBlockSize> read_buffer{};
+    while (true) {
+        size_t num_bytes_read{};
+        ErrorCode const error_code
+                = reader.try_read(read_buffer.data(), cReadBlockSize, num_bytes_read);
+        if (ErrorCode_EndOfFile == error_code) {
+            break;
+        }
+        if (ErrorCode_Success != error_code) {
+            throw OperationFailed(error_code, __FILENAME__, __LINE__);
+        }
+        archive_writer.write(read_buffer.data(), num_bytes_read);
+    }
+}
+
+auto write_archive_files(
+        FileWriter& archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        std::vector<std::string> const& segment_ids
+) -> void {
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        std::filesystem::path static_archive_file_path
+                = multi_file_archive_path / static_archive_file_name;
+        write_archive_file(static_archive_file_path, archive_writer);
+    }
+
+    std::filesystem::path segment_dir_path = multi_file_archive_path / cSegmentsDirname;
+    for (auto const& segment_id : segment_ids) {
+        std::filesystem::path segment_path = segment_dir_path / segment_id;
+        write_archive_file(segment_path, archive_writer);
+    }
+}
+}  // namespace
+
+auto write_single_file_archive(
+        ArchiveMetadata const& multi_file_archive_metadata,
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> void {
+    FileWriter archive_writer;
+    std::filesystem::path single_file_archive_path
+            = multi_file_archive_path.string()
+              + std::string(single_file_archive::cUnstructuredSfaExtension);
+
+    if (std::filesystem::exists(single_file_archive_path)) {
+        throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
+    }
+
+    archive_writer.open(
+            single_file_archive_path.string(),
+            FileWriter::OpenMode::CREATE_FOR_WRITING
+    );
+
+    auto const segment_ids = get_segment_ids(next_segment_id);
+
+    auto const packed_metadata = pack_single_file_archive_metadata(
+            multi_file_archive_metadata,
+            get_file_infos(multi_file_archive_path, segment_ids),
```

Comment:
we can put `get_file_infos` inside pack_single_file_archive_metadata, right?

It just needs to take segment_ids as its second argument